src/libraries/Common/src/Interop/OSX/System.Security.Cryptography.Native.Apple/Interop.Aead.cs

```diff
@@ -128,13 +170,31 @@ internal static unsafe void AesGcmEncrypt(
             fixed (byte* tagPtr = tag)
             fixed (byte* aadPtr = &GetSwiftRef(aad))
             {
-                AppleCryptoNative_AesGcmEncrypt(
-                    new UnsafeBufferPointer<byte>(keyPtr, key.Length),
-                    new UnsafeBufferPointer<byte>(noncePtr, nonce.Length),
-                    new UnsafeBufferPointer<byte>(plaintextPtr, plaintext.Length),
-                    new UnsafeMutableBufferPointer<byte>(ciphertextPtr, ciphertext.Length),
-                    new UnsafeMutableBufferPointer<byte>(tagPtr, tag.Length),
-                    new UnsafeBufferPointer<byte>(aadPtr, aad.Length),
+                var symmetricKeyData = new Data(keyPtr, key.Length);
+                var swiftKey = new SymmetricKey(symmetricKeyData);
+
+                var nonceData = new Data(noncePtr, nonce.Length);
+                var swiftNonce = new AesGcm.Nonce(nonceData);
+
+                var plaintextData = new Data(plaintextPtr, plaintext.Length);
+                var aadData = new Data(aadPtr, aad.Length);
+
+                void* messagePtr = &plaintextData;
+                void* authenticatedDataPtr = &aadData;
+
+                var sealedBox = new AesGcm.SealedBox();
+                SwiftIndirectResult swiftIndirectResult = new SwiftIndirectResult(sealedBox.payload);
+
+                CryptoKit.PInvoke_AesGcm_Seal(
+                    swiftIndirectResult,
+                    messagePtr,
+                    swiftKey.payload,
+                    swiftNonce.payload,
+                    authenticatedDataPtr,
+                    plaintextData.Metadata(),
+                    aadData.Metadata(),
+                    default(DataProtocol).WitnessTable(plaintextData),
+                    default(DataProtocol).WitnessTable(aadData),
```

Comment:
Could we define a wrapper of the P/Invokes that takes two `ISwiftType` values and calls the WitnessTable functions in the wrapper so we don't have to do it manually?

Reply:
I've moved Swift-specific features to the bindings to simulate projection tooling behavior. I believe it's reasonable to handle marshalling in the projections rather than in the user-facing APIs. Here's an example of the API:
```csharp
Data symmetricKeyData = new Data(keyPtr, key.Length);
SymmetricKey symmetricKey = new SymmetricKey(symmetricKeyData);

Data nonceData = new Data(noncePtr, nonce.Length);
ChaChaPoly.Nonce chaChaPolyNonce = new ChaChaPoly.Nonce(nonceData);

Data plaintextData = new Data(plaintextPtr, plaintext.Length);
Data aadData = new Data(aadPtr, aad.Length);

ChaChaPoly.SealedBox sealedBox = ChaChaPoly.seal(
    plaintextData,
    symmetricKey,
    chaChaPolyNonce,
    aadData,
    out SwiftError error);

if (error.Value != null)
{
    CryptographicOperations.ZeroMemory(ciphertext);
    CryptographicOperations.ZeroMemory(tag);
    throw new CryptographicException();
}

Data resultCiphertext = sealedBox.Ciphertext;
Data resultTag = sealedBox.Tag;

resultCiphertext.CopyBytes(ciphertextPtr, resultCiphertext.Count);
resultTag.CopyBytes(tagPtr, resultTag.Count);
```

Do you think the P/Invoke wrapper would still be beneficial in this case?